platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/ShadowGraph.java

```diff
@@ -108,39 +116,65 @@ public class ShadowGraph implements Clearable {
     /** the number of nodes in the network, used for debugging */
     private final int numberOfNodes;
 
+    private final NodeId selfId;
+
     /**
-     * Constructs a new instance.
+     * Tracks tipsets of events. Useful for efficiently determining whether one event is an ancestor of another.
      */
-    public ShadowGraph(final SyncMetrics syncMetrics) {
-        this(syncMetrics, -1);
-    }
+    private final TipsetTracker tipsetTracker;
+
+    /**
+     * The tipset of the latest self event, or null if there have been no self events.
+     */
+    private Tipset latestSelfEventTipset;
```

Comment:
This complicates the shadowgraph. Should we maybe go with a wrapper or extension that has a tipset?

Reply:
Per our discussion, I've removed this from `ShadowGraph.java` and maintain the data structure in an adjacent location.